### PR TITLE
Always pre-compute connector type during mapping resolving [HZ-2324]

### PR DESCRIPTION
**Pull request by Fly-Style**
Closes https://github.com/hazelcast/hazelcast/issues/24337

This patch introduces pre-computing of connector type during mapping resolving. Previously, if data connection was created, mapping was created with given data conn, and then data connection was removed, for each next query schema computation was broken (mapping can't find removed data conn -> fail). Good example is in #24337

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/parse/SqlCreateMapping.java [hunk 1]
```diff
@@ -84,8 +84,6 @@ public SqlCreateMapping(
         this.connectorType = connectorType;
         this.objectType = objectType;
         this.options = requireNonNull(options, "Options should not be null");
-
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/parse/SqlCreateMapping.java:87]: why did you remove that?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java [hunk 2]
```diff
@@ -159,7 +151,8 @@ private Mapping resolveMapping(Mapping mapping) {
         return new Mapping(
                 mapping.name(),
                 mapping.externalName(),
-                mapping.dataConnection(), type,
+                mapping.dataConnection(),
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: what will happen when someone changes DATA CONNECTION type, eg. from JDBC to Kafka? Will we be able to detect it in a meaningful way?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java [hunk 3]
```diff
@@ -311,6 +312,33 @@ public void when_mappingIsDeclaredWithDataConnection_then_itIsAvailable() throws
         assertThat(object).isEqualTo(name);
     }
 
+    /**
+     * Source : <a href=https://github.com/hazelcast/hazelcast/issues/24337">issue #24337</a>.
+     */
+    @Test
+    public void given_mappingIsDeclaredWithDataConn_when_DataConnWasRemoved_then_success() throws Exception {
+        // given
+        String dlName = randomName();
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java:321]: nit: `dcName`

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java [hunk 4]
```diff
@@ -311,6 +312,33 @@ public void when_mappingIsDeclaredWithDataConnection_then_itIsAvailable() throws
...
+        Map<String, String> options = new HashMap<>();
+        options.put("jdbcUrl", dbConnectionUrl);
+
+        // when
+        createDataConnection(instance(), dlName, "JDBC", false, options);
+        createJdbcMappingUsingDataConnection(name, dlName);
+        sqlService.execute("DROP DATA CONNECTION " + dlName);
+
+        // then
+        assertThat(dlName).isNotEqualTo(TEST_DATABASE_REF);
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java:333]: `dlName = randomName()` so that assertion does not make much sense

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java [hunk 5]
```diff
@@ -311,6 +312,33 @@ public void when_mappingIsDeclaredWithDataConnection_then_itIsAvailable() throws
...
+        assertRowsAnyOrder("SHOW DATA CONNECTIONS ", singletonList(new Row(TEST_DATABASE_REF)));
+
+        // Mapping doesn't provide data, since data connection was removed.
+        assertThatThrownBy(() -> sqlService.execute("SELECT * FROM " + name))
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java:337]: I think you should execute an unrelated SQL query or `SHOW MAPPINGS` to ensure that everything works and fully reproduce the error

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java (hunk 2 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: TDD definition is pretty clear for this and similar cases :

> When a query which uses a mapping based on a data connection is submitted, the current
settings of the data connection are used, and subsequent changes aren’t reflected, nor
do they change the query failure.

Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: are you sure that it is related to stored connections metadata or connection/mapping currently being used? AFAIR the intention was that when a _running_ processor uses a mapping/data connection and this object is altered, such processor continues to use version with which it had started. But there was no intention to make `data connection` only a template for creating mappings which afterwards are independent. For example, you can change password in `data connection` and all subsequently started jobs should use new password.

Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: > are you sure that it is related to stored connections metadata or connection/mapping currently being used?

From the wording used in TDD, I can make an assumption that it is related both to metadata and currently running jobs.

Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: > For example, you can change the password in the data connection and all subsequently started jobs should use new password.

Yes, and they will. I don't see any contradictions for this case - new job will use new data connection with updated password

Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: password - yes, but connectionType - no.

After this PR I can immediately file the following issue:
> I changed data connection (DROP+CREATE) from MySQL to Postgres, but my mappings created before the change still use MySQL dialect.

Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TableResolverImpl.java:154]: The wording in the TDD is unclear, I clarified this with Viliam what it was supposed to mean and we agreed that it is that running queries are not affected, but new queries from the mapping take the new settings. The password update is good example, another would be host update in jdbc url.

If you use the data connection in a job, the new settings is used even when a job is restarted.

**Review by frant-hartm — APPROVED** [review]

**Review by ivanthescientist — APPROVED** [review]
